Upgrade to Hazelcast 4

tsegismont opened the pull request:
Closes #118

tsegismont commented:
Hi @vbekiaris , the upgrade looks good except for locks. It seems that in CP unsafe mode, semaphore permits are not released when a node leaves. Does that ring a bell?

vbekiaris commented:
Hi @tsegismont! There is a known issue with cleanup in CP unsafe mode: https://github.com/hazelcast/hazelcast/issues/16932. Essentially the semaphore will be released when the CP unsafe session times out (default session timeout is 5 minutes). For example the test `io.vertx.it.litemembers.HazelcastClusteredAsynchronousLockTest#testLockReleasedForClosedNode` that fails due to this issue, passes with this additional config:

```
Config config = ConfigUtil.loadConfig();
    config.getCPSubsystemConfig().setSessionHeartbeatIntervalSeconds(1)
          .setSessionTimeToLiveSeconds(5);
```

tsegismont commented:
@vbekiaris thanks for the pointer! So if I understand correctly, we can make these tests pass by reducing the heartbeat timeout?

Then we can document this behavior in Vert.x 4 for our users. Those who stick to unsafe mode (the default in this cluster manager) would see this behavior, those who configure HZ CP subsytem properly wouldn't.

vbekiaris commented:
@tsegismont you are right. Notice that the CP subsystem config has a separate CP session heartbeat (just clarifying to avoid confusion with cluster membership [failure detection heartbeat](https://docs.hazelcast.org/docs/4.0.1/manual/html-single/index.html#deadline-failure-detector)).

[Fix PRs](https://github.com/hazelcast/hazelcast/pull/17029) have been prepared and could probably get merged in time for release 4.0.2. I will update this thread once merged.

tsegismont commented:
Great news @vbekiaris ! In the meantime I updated the test config (961ec6a) and fixed a CL issue (485b783). The test suite is all green so I believe the HZ 4 upgrade will make it to Vert.x 4.0 :+1:

vietj commented on src/main/java/io/vertx/spi/cluster/hazelcast/impl/HazelcastAsyncMap.java:
```diff
@@ -69,26 +68,26 @@ public HazelcastAsyncMap(VertxInternal vertx, IMap<K, V> map) {
-      map.set(kk, HazelcastServerID.convertServerID(vv), ttl, TimeUnit.MILLISECONDS);
-      fut.complete();
-    }, false);
+    ContextInternal context = vertx.getOrCreateContext();
```
why is executeBlocking disappearing here ?

tsegismont replied:
Because all `IMap` now implement the HZ async APIs and we no longer need two implementations.

vbekiaris commented:
@tsegismont fyi Hazelcast IMDG 4.0.2 was just released, including the fix for https://github.com/hazelcast/hazelcast/issues/16932

tsegismont commented:
Great! Thanks for the heads-up Vassilis!
